### PR TITLE
Check for PVC and DataVolume existence in a central place

### PR DESCRIPTION
**Pull request by rmohr**
Only proceed to template rendering when all PVCs are present.  The template renderer also checks for PVC existence, but if the PVC appeared between the check and the template rendering, the limited render-check will miss subtleties like checking if the PVC belongs to a DV and an import needs to be finished before creating the launcher pod.

This fixes a race, where sometimes a launcher pod for a VMI could already be created while the DV was in the middle of creating the PVC and importing.

As a consequence the VMI controller thinks that the import is already done and will just proceed with the VM start, which leads to a broken import and a failing VMI start.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Ensure existence of all PVCs attached to the VMI before creating the VM target pod.
```

### CODE REVIEW

**Comment by rmohr** [comment]
We have a "flaky" setup where CDI sometims fails over on leader election because of network latency. Due to the general unpredictable network latency we hit there quite often the case where the disk gets corrupted, so a very real issue.

File: pkg/virt-controller/watch/vmi.go [hunk 1]
```diff
@@ -1228,12 +1228,14 @@ func (c *VMIController) handleSyncDataVolumes(vmi *virtv1.VirtualMachineInstance
 		if volume.VolumeSource.DataVolume != nil || volume.VolumeSource.PersistentVolumeClaim != nil {
 			volumeReady, volumeWffc, err := storagetypes.VolumeReadyToAttachToNode(vmi.Namespace, volume, dataVolumes, c.dataVolumeInformer, c.pvcInformer)
 			if err != nil {
-				// Keep existing behavior of missing PVC = ready. This in turn triggers template render, which sets conditions and events, and fails appropriately
 				if _, ok := err.(storagetypes.PvcNotFoundError); ok {
-					continue
+					// due to the eventually consistent nature of controllers, CDI or users may need some time to actually crate the PVC.
```
Comment (awels) [inline pkg/virt-controller/watch/vmi.go:1232]: typo `crate` should be `create`

**Review by awels — COMMENTED** [review]
Okay I see the problem, we start with `ready := true` and if all the volumes are skipped due to the PVCs not being there, it says, yay we are ready, while we are not. This PR looks fine to me, I think we can also just set `ready = false` right before the continue, but there is really no reason to continue looping if we know it will never be ready.

**Comment by rmohr** [comment]
> Okay I see the problem, we start with `ready := true` and if all the volumes are skipped due to the PVCs not being there, it says, yay we are ready, while we are not. This PR looks fine to me, I think we can also just set `ready = false` right before the continue, but there is really no reason to continue looping if we know it will never be ready.

Yes, we could also just return not-ready. I just wanted to keep for now the known semantic (the template-renderer would throw such a sync error if the PVC does not exist -- but the race above can lead to false discovery there). I think it would make sense to long-term change this and treat this as a non-error situation where we indicate on a condition that we are waiting for the volume.

**Comment by awels** [comment]
looks like some test failures due to this change.

**Comment by mhenriks** [comment]
> > Okay I see the problem, we start with `ready := true` and if all the volumes are skipped due to the PVCs not being there, it says, yay we are ready, while we are not. This PR looks fine to me, I think we can also just set `ready = false` right before the continue, but there is really no reason to continue looping if we know it will never be ready.
> 
> Yes, we could also just return not-ready. I just wanted to keep for now the known semantic (the template-renderer would throw such a sync error if the PVC does not exist -- but the race above can lead to false discovery there). I think it would make sense to long-term change this and treat this as a non-error situation where we indicate on a condition that we are waiting for the volume.

I think this change is fine but my biggest takeaway in rereading this is that `handleSyncDataVolumes` really shouldn't return an error (and neither should `VolumeReadyToAttachToNode` but that's not as important).  The controller is watching PVCs and should reconcile when the PVC comes into existence.  So no need to rely on controller error handling (exponential retry).

**Comment by rmohr** [comment]
> looks like some test failures due to this change.

Fixed (:crossed_fingers:).

**Comment by rmohr** [comment]
> I think this change is fine but my biggest takeaway in rereading this is that `handleSyncDataVolumes` really shouldn't return an error (and neither should `VolumeReadyToAttachToNode` but that's not as important). The controller is watching PVCs and should reconcile when the PVC comes into existence. So no need to rely on controller error handling (exponential retry).

Addressed.

**Review by mhenriks — COMMENTED** [review]
Looks good, but I just noticed that the pvc informer does not have an `AddFunc`.  May be needed for this case.

https://github.com/rmohr/kubevirt/blob/datavolume-race/pkg/virt-controller/watch/vmi.go#L185-L187

**Comment by rmohr** [comment]
> Looks good, but I just noticed that the pvc informer does not have an `AddFunc`. May be needed for this case.
> 
> https://github.com/rmohr/kubevirt/blob/datavolume-race/pkg/virt-controller/watch/vmi.go#L185-L187

Hm, if the PVC is not in the cache, it would trigger an add as soon as it appears in the cache, or do you have a specific corner-case in mind?

**Comment by mhenriks** [comment]
> Hm, if the PVC is not in the cache, it would trigger an add as soon as it appears in the cache, or do you have a specific corner-case in mind?

One of the goals of this PR is to trigger reconcile when the PVC is created.  PVC creation will trigger the `AddFunc` callback.  Which we are not handling (so will not reconcile).  BUT I suspect that the PVC will be updated soon after it is created and that will trigger the `UpdateFunc` callback which we ARE handling.

I don't think it will matter at the end of the day, but I think it will be wise to add a `AddFunc` callback on the PVC informer

**Comment by rmohr** [comment]
> I don't think it will matter at the end of the day, but I think it will be wise to add a `AddFunc` callback on the PVC informer

You are completely right. I thought it wast he other way, that we only have `AddFunc` and not `UpdateFunc`. Will add it.

**Pull request merged**